26/WAKU-PAYLOAD (#413)
vacp2p/rfc: content/docs/rfcs/26/README.md

---
slug: 26
title: 26/WAKU-PAYLOAD
name: Waku Message Payload Encryption
status: draft
editor: Oskar Thoren <oskar@status.im>
contributors:
---

> This specification describes how Waku provides provides confidentiality, authenticity, and integrity, as well as some form unlinkability.

Reviewer: ```suggestion
This specification describes how Waku provides confidentiality, authenticity, and integrity, as well as some form of unlinkability.
```

Specifically, it describes how encryption, decryption and signing works in [6/WAKU1](/spec/6) and in [10/WAKU2](/spec/10) with [14/WAKU-MESSAGE version 1](/spec/14/#version1).

> This specification effectively replaces [7/WAKU-DATA](/spec/7) as well as [6/WAKU1 Payload encryption](/spec/6/#payload-encryption) but written in a way that is agnostic and self-contained for Waku v1 and Waku v2.

Reviewer: Does that mean in the future we are going to remove those specs entirely? Or will they be accessible? I am asking to know what level of details should be expected in the current specs

Author: The stable ones should remain stable and not materially change until they are deprecated. Status client currently depends on both of these.

Large sections of the specification originate from [EIP-627: Whisper spec](https://eips.ethereum.org/EIPS/eip-627) as well from [RLPx Transport Protocol spec (ECIES encryption)](https://github.com/ethereum/devp2p/blob/master/rlpx.md#ecies-encryption) with some modifications.

## Design requirements

- *Confidentiality*: The adversary should not be able to learn what data is being sent from one Waku node to one or several other Waku nodes.
- *Authenticity*: The adversary should not be able to cause Waku endpoint to accept data from any third party as though it came from the other endpoint.
- *Integrity*: The adversary should not be able to cause a Waku endpoint to accept data that has been tampered with.

Notable, *forward secrecy* is not provided for at this layer.
If this property is desired, a more fully featured secure communication protocol can be used on top, such as [Status 5/SECURE-TRANSPORT](https://specs.status.im/spec/5).

> It also provides some form of *unlinkability* since:
> - only participants who are able to decrypt a message can see its signature
> - payload are padded to a fixed length

Reviewer: ```suggestion
### Additional features
- *Message Unlinkability*: Waku messages are not attributable to their authors unless by the conversation participants. 
The message unlinkability is provided because:
- Messages are first signed and then encrypted.  As such, the identity of the message signer remains anonymous and unlikable to the message against anyone who only has access to the ciphertext. Only participants trusted by the message author and with the corresponding decryption key would be able to open the ciphertext, obtain the signature and learn the author of the message. 
- Payloads are padded to a fixed length
```

Reviewer: I have put my understanding about the unlinkability in the suggested changes, please feel free to modify it as it fits. :)
Also, I think it would be good to elaborate on the relation between the padded payload and the achieved unlinkability

> ## Cryptographic primitives

Reviewer: I'd provide a link for each one of these primitives

Author: As discussed, not obvious where to refer. We can do this in a follow up PR though if we find some good canonical links.

> - AES-256-GCM

Reviewer 2: AES-128-CTR is actually used for ECIES as written below so it's a bit weird to have AES mentioned twice in the list. Maybe add a bit of context here? E.g. "AES-256-GCM (symmetric encryption)".

Author: added

- ECIES

> - ECSDA

Reviewer: ```suggestion
- ECDSA
```

> - ECSDA

Author: ```suggestion
- ECDSA
```

- KECCAK-256

ECIES is using the following cryptosystem:
- Curve: secp256k1

> - KDF: NIST SP 800-56 Concatenation Key Derivation Function

Reviewer 2: I think we should specify that sha256 is used. Looking at NIST SP 800-56 it provides a list of approved hash function but does not specify it which can lead to ambiguity (I have seen libraries using SHA-512 for example)

Author: It is mentioned in the line below `- MAC: HMAC with SHA-256`

Author: made explicit for KDF too

- MAC: HMAC with SHA-256
- AES: AES-128-CTR

## Specification

> For 6/WAKU1, the `data` field is used in the `waku envelope`, and the field MAY contain the encrypted payload.

Reviewer: If this spec is going to be stand-alone, then I think the protbuf of waku envelope would be needed

Author: There is no protobuf in a waku envelope and Waku 1 is stable so the data format isn't going to change.

For version 1 we are keeping compatibility with Waku v1 to make core integration easier. For general improvements of this nature, I suggest we capture it as requirement here for version 2 https://github.com/vacp2p/rfc/issues/414

For 10/WAKU2, the `payload` field is used in `WakuMessage` and MAY contain the encrypted payload.

The fields that are concatenated and encrypted as part of the `data`/`payload` field are:
 - flags
 - payload length

>  - payload

Reviewer: ```suggestion
 - payload/data length
 - payload/data
```

Author: oops, this is wrong, shouldn't have merged, reverting

 - padding
 - signature

### ABNF

> Using [Augmented Backus-Naur form (ABNF)](https://tools.ietf.org/html/rfc5234) we have the following format:

Reviewer: ```suggestion
The `payload` and `data` fields are structured based on the [Augmented Backus-Naur form (ABNF)](https://tools.ietf.org/html/rfc5234)  as given below:
```

Author: ABNF is a general purpose tool, the existing phrasing is more accurate

```abnf

> ; 1 byte; first two bits contain the size of auxiliary field, 

Reviewer: What is the auxiliary field in here?

Reviewer 3: Seems like auxiliary field was rename `payload-length`. So this should be updated

Author: updated

; third bit indicates whether the signature is present.
flags           = 1OCTET

; contains the size of payload.
payload-length  = 4*OCTET

; byte array of arbitrary size (may be zero).
payload         = *OCTET

; byte array of arbitrary size (may be zero).
padding         = *OCTET

; 65 bytes, if present.
signature       = 65OCTET

data            = flags payload-length payload padding [signature]
```

> ### Asymmetric encryption

Reviewer: I suggest having a subsection ## Encryption and then under that briefly explain that the types of encryptions that are available.

Asymmetric encryption uses the standard Elliptic Curve Integrated Encryption Scheme (ECIES) with SECP-256k1 public key.

> For more details, see the section below on ECIES encryption.

Reviewer: Why not explaining ECIES as a subsection here?
I'd reorder these subsections as
### Signature
### Encryption
#### Symmetric
#### Asymmetric
##### ECIES

Author: done

### Symmetric encryption

> Symmetric encryption uses AES-256-GCM for [authenticated encryption](https://en.wikipedia.org/wiki/Authenticated_encryption), with a 16 byte authentication tag and a 12 byte IV (nonce).

Reviewer: My comment is about line 88, but for some reason GH does not le tme place my comment on that line!
Is `ciphertext` introduced in the specs? could not find it

Reviewer: ```suggestion
Symmetric encryption uses AES-256-GCM for [authenticated encryption](https://en.wikipedia.org/wiki/Authenticated_encryption). 
The output of encryption is of the form (`ciphertext`, `tag`, `iv`) where `ciphertext` is the encrypted message, `tag` is a 16 byte message authentication tag and `iv` is a 12 byte initialization vector  (nonce). 
```

The message authentication `tag` and initialization vector `iv` field MUST be appended to the resulting `ciphertext`, in that order.
Note that previous specifications and some implementations might refer to `iv` as `nonce` or `salt`.

### Signature

Those unable to decrypt the envelope data are also unable to access the signature.

> The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity.

Reviewer: what is originator identity? is it the originator signature verification key? does it correspond to something specific in waku e.g., libp2p peer Id?

Author: up to client, agnostic. Whoever is sending the messaging and signing it and has access to private key

The signature is serialized as the concatenation of the `r`, `s` and `v` parameters of the SECP-256k1 ECDSA signature, in that order.

> `r` and `s` MUST be big-endian encoded, fixed-width 256-bit unsigned.

Reviewer: Do `r` and `s` refer to other available notations in other specs? how should they be interpreted?

Author: no canonical reference afaict, more idiomatic. Added reference to Wikipedia which uses same idiom.

Author: added ref to yellow paper

> `v` MUST be an 8-bit big-endian encoded, non-normalized and should be either 27 or 28.

Reviewer: Ditto `v`

Author: added ref to yellow paper

### Padding

The padding field is used to align data size, since data size alone might reveal important metainformation.
Padding can be arbitrary size.

> However, it is recommended that the size of Data Field (excluding the IV and tag) before encryption (i.e. plain text) SHOULD be a multiple of 256 bytes.

Reviewer: What is Data Field? does it refer to the data field in the waku envelope?

Reviewer: The IV and tag are related to the authenticated encryption, right? the way it is stated is like that the authenticated encryption should be used or is assumed to be used.

### ECIES encryption

> 

Reviewer: Wondering why ECIES is explained in detail? but for example, the ECDSA is explained just at the interface level

Author: discussed in call; ECIES is a collection of primitives, whereas ECSDA is a primitive

This section originates from the [RLPx Transport Protocol spec](https://github.com/ethereum/devp2p/blob/master/rlpx.md#ecies-encryption) spec with minor modifications.

The cryptosystem used is:

- The elliptic curve secp256k1 with generator `G`.
- `KDF(k, len)`: the NIST SP 800-56 Concatenation Key Derivation Function.
- `MAC(k, m)`: HMAC using the SHA-256 hash function.
- `AES(k, iv, m)`: the AES-128 encryption function in CTR mode.

> Special notation used: `X || Y` denotes concatenation of X and Y.

Reviewer: ```suggestion
Special notation used: `X || Y` denotes concatenation of `X` and `Y`.
```

> Alice wants to send an encrypted message that can be decrypted by Bobs static private key `kB`. Alice knows about Bobs static public key `KB`.

Reviewer: ```suggestion
Alice wants to send an encrypted message that can be decrypted by Bob's static private key `kB`. Alice knows about Bobs static public key `KB`.
```

To encrypt the message `m`, Alice generates a random number `r` and corresponding elliptic curve public key `R = r * G` and computes the shared secret `S = Px` where `(Px, Py) = r * KB`.
She derives key material for encryption and authentication as `kE || kM = KDF(S, 32)` as well as a random initialization vector `iv`.
Alice sends the encrypted message `R || iv || c || d` where `c = AES(kE, iv , m)` and `d = MAC(sha256(kM), iv || c)` to Bob.

For Bob to decrypt the message `R || iv || c || d`, he derives the shared secret `S = Px` where `(Px, Py) = kB * R` as well as the encryption and authentication keys `kE || kM = KDF(S, 32)`.
Bob verifies the authenticity of the message by checking whether `d == MAC(sha256(kM), iv || c)` then obtains the plaintext as `m = AES(kE, iv || c)`.

### Decoding a message

> In order to decode a message, a node SHOULD try to apply both symmetric and asymmetric decryption operations.

Reviewer: Is it because the type of encryption is not included in the message?

Reviewer 3: Yes

Author: added clarification

## Design requirements

- *Confidentiality*: The adversary should not be able to learn what data is being sent from one Waku node to one or several other Waku nodes.
- *Authenticity*: The adversary should not be able to cause Waku endpoint to accept data from any third party as though it came from the other endpoint.
- *Integrity*: The adversary should not be able to cause a Waku endpoint to accept data that has been tampered with.

Notable, *forward secrecy* is not provided for at this layer.
If this property is desired, a more fully featured secure communication protocol can be used on top, such as [Status 5/SECURE-TRANSPORT](https://specs.status.im/spec/5).

It also provides some form of *unlinkability* since:
- only participants who are able to decrypt a message can see its signature
- payload are padded to a fixed length

## Cryptographic primitives

- AES-256-GCM
- ECIES
- ECSDA
- KECCAK-256

ECIES is using the following cryptosystem:
- Curve: secp256k1
- KDF: NIST SP 800-56 Concatenation Key Derivation Function
- MAC: HMAC with SHA-256
- AES: AES-128-CTR

## Specification

> For 6/WAKU1, the `data` field is used in the `waku envelope`, and the field MUST contain the encrypted payload.

Author: ```suggestion
For 6/WAKU1, the `data` field is used in the `waku envelope`, and the field SHOULD contain the encrypted payload.
```

> For 10/WAKU2, the `payload` field is used in `WakuMessage` and MUST contain the encrypted payload.

Reviewer 4: Since this replaces https://rfc.vac.dev/spec/14/#payload-encryption, it may get lost that payload encryption remains optional for `10/WAKU2`.

Author: True. I actually think we could make it optional here too, that's the original behavior.

Author: ```suggestion
For 10/WAKU2, the `payload` field is used in `WakuMessage` and SHOULD contain the encrypted payload.
```

Author: @jm-clius changing to a SHOULD, wdyt?

Reviewer 4: LGTM

Reviewer 3: Sorry for the nitpick, but I believe `MAY` is the right term for options?

Author: MAY, fixed

> The fields that are concatenated and encrypted as part of the field are:

Author: ```suggestion
The fields that are concatenated and encrypted as part of the `data`/`payload` field are:
```

 - flags

>  - auxiliary field

Author: ```suggestion
 - payload length
```

 - payload
 - padding
 - signature

### ABNF

Using [Augmented Backus-Naur form (ABNF)](https://tools.ietf.org/html/rfc5234) we have the following format:

```abnf
; 1 byte; first two bits contain the size of auxiliary field, 
; third bit indicates whether the signature is present.
flags           = 1OCTET

; contains the size of payload.
payload-length  = 4*OCTET

; byte array of arbitrary size (may be zero).
payload         = *OCTET

; byte array of arbitrary size (may be zero).
padding         = *OCTET

; 65 bytes, if present.
signature       = 65OCTET

; 12 bytes, if present (in case of symmetric encryption).

> iv              = 12OCTET

Reviewer 3: iv goes last

Author: After discussion removing this

; 16 bytes, if present (in case of symmetric encryption).
tag             = 16OCTET

> data            = flags auxiliary-field payload padding [signature]

Reviewer 2: I believe this is changed now

Author: From before? correct yeah, it *could* be wrong but from what I can tell in the nim-eth / geth code it was largely confusing, because it made "salt" seem special, as opposed to integrated with AES-GCM appended tag+iv. Ditto for asymmetric case. Removing it seems to make the spec and behavior cleaner, assuming it is accurate.

Reviewer 2: I mean

```suggestion
data            = flags payload-length payload padding [signature] [iv] [tag]
```

Author: That is confusing because it doesn't represent the encrypted part. It also doesn't capture a similar thing happening to asymmetric encryption... please re-read this and related parts.

Author: The alternative would be to pull out encrypted field and use that, in which case it'd be something like

`encrypted = cipher iv tag`

I didn't do this because

1) It is standard for these primitives and mentioned
2) It'd require doing it for ECIES too, and seems like it'd be redundant/error prone; I'd rather defer such details to other specs, and every implementer should use an existing library for this. We aren't trying to reinvent the wheel here.

Reviewer 2: Sorry, I realize I am trying to second guess things instead of just pointing out what I don't understand:
- what is `data` here?
- what is `auxiliary-field?`

Author: Oh right, auxiliary-field should indeed be payload-length, forgot to change that. `data` is what is being encrypted, that can probably be clarified above. Thanks!

Author: ```suggestion
data            = flags payload-length payload padding [signature]
```

> data            = flags payload-length payload padding [signature]

Reviewer 3: optional iv is missing: [iv]

Reviewer 3: and tag is missing also. 

Is there no way to specify encrypted data (or functions in general) in ABNF by the way?

Something like:
```abnf 
encrypted = encrypt( flags payload-length payload padding [signature] )

; tag & iv when aesgcm encryption
data = encrypted [tag] [iv]

Author: Alt

```
plaintext =  flags payload-length payload padding [signature]

; ciphertext = encrypt(plaintext)

; tag & iv when aesgcm
encrypted_data = cipher [tag] [iv]
```

EDIT: After discussion probably more clear to leave as is, with ABNF describes plain text, then you should apply aesgcm or ecies, and those sections should give you enough info to do it correctly, but it is more of an "inner" level concern, nstead of half-assing it with only adding "salt" like whisper spec did

Author: removing from abnf as implementation detail

```

### Asymmetric encryption

Asymmetric encryption uses the standard Elliptic Curve Integrated Encryption Scheme (ECIES) with SECP-256k1 public key.
For more details, see the section below on ECIES encryption.

### Symmetric encryption

Symmetric encryption uses AES-256-GCM for [authenticated encryption](https://en.wikipedia.org/wiki/Authenticated_encryption), with a 16 byte authentication tag and a 12 byte IV (nonce).

> The message authentication `tag` and initialization vector `iv` field MUST be appended to the resulting `ciphertext`, in that order.

Reviewer 3: I guess in the original specification the tag was not mentioned as it is a MUST to have integrity and authenticity in AES-GCM. 
But I think it is indeed good to mention it here explicitly, also as I'm not sure if it is specified as how it should be transferred (append, prefix, whatever..).

Note that previous specifications and some implementations might refer to `iv` as `nonce` or `salt`.

### Signature

Those unable to decrypt the envelope data are also unable to access the signature.
The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity.
The signature is serialized as the concatenation of the `r`, `s` and `v` parameters of the SECP-256k1 ECDSA signature, in that order.
`r` and `s` MUST be big-endian encoded, fixed-width 256-bit unsigned.
`v` MUST be an 8-bit big-endian encoded, non-normalized and should be either 27 or 28.

### Padding

The padding field is used to align data size, since data size alone might reveal important metainformation.
Padding can be arbitrary size.

> However, it is recommended that the size of Data Field (excluding the IV) before encryption (i.e. plain text) SHOULD be factor of 256 bytes.

Author: I'm not 100% sure if this is accurate? Specifically the "excluding the IV" part

Reviewer 3: Yes, the padding is added before encryption. So not only iv but also tag is excluded. Rather weird that it was written like that, I think it is more clear to just say that the padding is done on the plain text before encryption.

So, `flags + payload-length +  payload + padding [+ signature]` is a multiple of 256. The term factor of 256 bytes is also wrong here.

Author: ```suggestion
However, it is recommended that the size of Data Field (excluding the IV and tag) before encryption (i.e. plain text) SHOULD be factor of 256 bytes.
```

> However, it is recommended that the size of Data Field (excluding the IV) before encryption (i.e. plain text) SHOULD be factor of 256 bytes.

Author: ```suggestion
However, it is recommended that the size of Data Field (excluding the IV) before encryption (i.e. plain text) SHOULD be mulitple of 256 bytes.
```

> However, it is recommended that the size of Data Field (excluding the IV and tag) before encryption (i.e. plain text) SHOULD be factor of 256 bytes.

Author: ```suggestion
However, it is recommended that the size of Data Field (excluding the IV and tag) before encryption (i.e. plain text) SHOULD be multiple of 256 bytes.
```

---
slug: 26
title: 26/WAKU-PAYLOAD
name: Waku Message Payload Encryption
status: draft
editor: Oskar Thoren <oskar@status.im>
contributors:
---

This specification describes how encryption, decryption and signing works in
[6/WAKU1](/spec/6) and in [10/WAKU2](/spec/10) with [14/WAKU-MESSAGE version
1](/spec/14/#version1).

It effectively replaces [7/WAKU-DATA](/spec/7) as well as [6/WAKU1 Payload
encryption](/spec/6/#payload-encryption) but written in a way that is agnostic
and self-contained for Waku v1 and Waku v2.

## Cryptographic primitives

- AES-256-GCM
- ECIES
- ECSDA
- KECCAK-256

## Payload encryption

Asymmetric encryption uses the standard Elliptic Curve Integrated Encryption Scheme (ECIES) with SECP-256k1 public key.

> In case of a signature being provided, the public key is recoverable from by utilizing the `v` parameter.

Reviewer 2: ```suggestion
In case of a signature being provided, the public key is recoverable by utilizing the `v` parameter.
```

Symmetric encryption uses AES-256-GCM for authenticated encryption [TODO], with a 16 byte authentication tag 16 and a 12 byte IV (nonce).

## Specification

For 6/WAKU1, the `data` field is used within the `waku envelope`, the field MUST contain the encrypted payload of the `waku envelope`.

For 10/WAKU2, the `payload` field is used within the `WakuMessage` and MUST contain the encrypted payload of `Waku Message`.

The fields that are concatenated and encrypted as part of the field are:
 - flags
 - auxiliary field
 - payload
 - padding
 - signature
 

> In case of symmetric encryption, a `salt`  (a.k.a. AES Nonce, 12 bytes) field MUST be appended. 

Reviewer 2: We are using 3 terminology (salt, iv, nonce) to refer to the same thing. Maybe best to reduce to 1 or at least 2? (e.g, IV + Nonce).

Author: Yeah agree, usually IV or nonce is used for this in literature afaict, somehow salt was used in early Whisper specs, which usually goes with passwords afaik. (Haven't gotten to this section yet)

### ABNF

Using [Augmented Backus-Naur form (ABNF)](https://tools.ietf.org/html/rfc5234) we have the following format:

> ```abnf

Reviewer 2: As mentioned in the call, needs to clarify where the 16 bytes authentication tag goes.

Also, when using ECIES I believe an ephemeral public key needs to be communicated, where does it go in the payload?

https://cryptobook.nakov.com/asymmetric-key-ciphers/ecies-public-key-encryption

Author: it should be clearer now, take a look

; 1 byte; first two bits contain the size of auxiliary field, 
; third bit indicates whether the signature is present.
flags           = 1OCTET

; contains the size of payload.
auxiliary-field = 4*OCTET

; byte array of arbitrary size (may be zero)
payload         = *OCTET

; byte array of arbitrary size (may be zero).
padding         = *OCTET

; 65 bytes, if present.
signature       = 65OCTET

> ; 2 bytes, if present (in case of symmetric encryption).
> salt            = 2OCTET

Reviewer 2: ```suggestion
; 12 bytes, if present (in case of symmetric encryption).
salt            = 12OCTET
```
